# Baseball Data Analysis


In [21]:
import pandas as pd
import numpy as np

##### The data for this project has been selected from http://seanlahman.com/download-baseball-database/

The goal of this project is to import statistical data from the game of baseball, clean said data to be able to use for our analysis, and perform manipulations of the data tables to make desired specific conclusions about certain points of interest as it pertains to the game.

In [22]:
df = pd.read_csv("Baseball Teams.csv")
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3015 entries, 0 to 3014
Data columns (total 48 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   yearID          3015 non-null   int64  
 1   lgID            2965 non-null   object 
 2   teamID          3015 non-null   object 
 3   franchID        3015 non-null   object 
 4   divID           1498 non-null   object 
 5   Rank            3015 non-null   int64  
 6   G               3015 non-null   int64  
 7   Ghome           2616 non-null   float64
 8   W               3015 non-null   int64  
 9   L               3015 non-null   int64  
 10  DivWin          1470 non-null   object 
 11  WCWin           834 non-null    object 
 12  LgWin           2987 non-null   object 
 13  WSWin           2658 non-null   object 
 14  R               3015 non-null   int64  
 15  AB              3015 non-null   int64  
 16  H               3015 non-null   int64  
 17  2B              3015 non-null   i

#### This master table contains 47 columns with a total of 3015 entries
#### Below are each column heading with descriptors of the data in each entry

yearID:         Year\
lgID:           League\
teamID:         Team\
franchID:       Franchise (links to TeamsFranchise table)\
divID:          Team's division\
Rank:           Position in final standings\
G:              Games played\
GHome:          Games played at home\
W:              Wins\
L:              Losses\
DivWin:         Division Winner (Y or N)\
WCWin:          Wild Card Winner (Y or N)\
LgWin:          League Champion(Y or N)\
WSWin:          World Series Winner (Y or N)\
R:              Runs scored\
AB:             At bats\
H:              Hits by batters\
2B:             Doubles\
3B:             Triples\
HR:             Homeruns by batters\
BB:             Walks by batters\
SO:             Strikeouts by batters\
SB:             Stolen bases\
CS:             Caught stealing\
HBP:            Batters hit by pitch\
SF:             Sacrifice flies\
RA:             Opponents runs scored\
ER:             Earned runs allowed\
ERA:            Earned run average\
CG:             Complete games\
SHO:            Shutouts\
SV:             Saves\
IPouts:         Outs Pitched (innings pitched x 3)\
HA:             Hits allowed\
HRA:            Homeruns allowed\
BBA:            Walks allowed\
SOA:            Strikeouts by pitchers\
E:              Errors\
DP:             Double Plays\
FP:             Fielding  percentage\
\
name:           Team's full name\
park:           Name of team's home ballpark\
attendance:     Home attendance total\
BPF:            Three-year park factor for batters\
PPF:            Three-year park factor for pitchers\
teamIDBR:       Team ID used by Baseball Reference website\
teamIDlahman45: Team ID used in Lahman database version 4.5\
teamIDretro:    Team ID used by Retrosheet\

In [68]:
print("We can also take a look at the first few rows of the table to better understand the set we are working with")
df.head(5)

We can also take a look at the first few rows of the table to better understand the set we are working with


yearID lgID teamID franchID divID  Rank   G  Ghome   W   L  ... attendance  \
0    1871  NaN    BS1      BNA   NaN     3  31    NaN  20  10  ...        NaN   
1    1871  NaN    CH1      CNA   NaN     2  28    NaN  19   9  ...        NaN   
2    1871  NaN    CL1      CFC   NaN     8  29    NaN  10  19  ...        NaN   
3    1871  NaN    FW1      KEK   NaN     7  19    NaN   7  12  ...        NaN   
4    1871  NaN    NY2      NNA   NaN     5  33    NaN  16  17  ...        NaN   

   BPF  PPF teamIDBR  teamIDlahman45  teamIDretro    OBP    SLG    OPS   WHIP  
0  103   98      BOS             BS1          BS1  0.434  0.502  0.936  1.482  
1  104  102      CHI             CH1          CH1  0.390  0.443  0.833  1.339  
2   96  100      CLE             CL1          CL1  0.368  0.460  0.828  1.571  
3  101  107      KEK             FW1          FW1  0.322  0.332  0.654  1.669  
4   90   88      NYU             NY2          NY2  0.357  0.396  0.753  1.416  

[5 rows x 52 columns]

With the current table of data, we as the analysts can come up with a few more measures by which to compare the teams. We will be creating composite statistcs that will take larger columns of data and boil them down to singular numbers by which we can judge temas for offensive and defensive output

In [24]:
#Addition of new statistics to the dataframe for our analysis

#onBase percentage (total amount of hits, doubles, triples, home runs and walks divided by the total number of at bats)
df["OBP"] = np.round((df.H +df['2B']+df["3B"]+df.HR + df.BB)/df.AB, 3)

#Slugging percentage (total number of bases accumulated by a batter divided by the total number of at bats)
df["SLG"] = np.round(((df.H)+(df["2B"] * 2)+ (df["3B"] *3)+(df.HR * 4))/ df.AB,3)

#OPS (Our on base percentage plus our slugging percentage gives a better picture of the statistical success\
#     of the teams cumulative batting statistics)
df["OPS"] = np.round(df.OBP + df.SLG,3)

#WHIP (Total walks and hits by the pitching staff divided by the amount of innings pitched,\
#      describes how many batters are reaching base any one given inning)
df["WHIP"] = np.round((df.HA + df.BBA)/ (df.IPouts/3),3)

Now to separate the data into what we need from the main table, lets select our columns of interest.
With this data we can then make our analysis on the selected statistical areas for each team

In [25]:
teamData = df[["name","yearID","lgID","teamID","franchID","divID","Rank","G","W","L","DivWin","WCWin","LgWin","WSWin","R","AB","H","2B","3B",\
               "HR","BB","OBP","SLG","OPS","SO","ER","ERA","WHIP","IPouts","HA","HRA","BBA","SOA","E","DP","FP"]]

To narrow this data even further down into more usable and meaningful numbers, 
we will only include the years of the MLB after the modern era of baseball began following World War I.

In [26]:
modernData = teamData[teamData.yearID >= 1920]
modernData.head()

name  yearID lgID teamID franchID divID  Rank    G   W   L  \
703     Boston Red Sox    1920   AL    BOS      BOS   NaN     5  154  72  81   
704    Brooklyn Robins    1920   NL    BRO      LAD   NaN     1  155  93  61   
705      Boston Braves    1920   NL    BSN      ATL   NaN     7  153  62  90   
706  Chicago White Sox    1920   AL    CHA      CHW   NaN     2  154  96  58   
707       Chicago Cubs    1920   NL    CHN      CHC   NaN     5  154  75  79   

     ...   ERA   WHIP IPouts    HA  HRA  BBA  SOA    E   DP     FP  
703  ...  3.82  1.392   4185  1481   39  461  481  183  131  0.972  
704  ...  2.62  1.197   4281  1381   25  327  553  226  118  0.966  
705  ...  3.54  1.356   4158  1464   39  415  368  239  125  0.964  
706  ...  3.59  1.351   4158  1467   45  405  438  194  142  0.969  
707  ...  3.27  1.326   4164  1459   37  382  508  225  112  0.965  

[5 rows x 36 columns]

In [67]:
#Calculating most wins and Best Team ever 

mostWins = modernData.W.max()
BestTeams = modernData[(modernData.W == mostWins)]
print("The team with the most wins in the history post WWI is:")
print(BestTeams)

print("\n")
#Most losses and worst team

mostLosses = modernData.L.max()
worstTeams = modernData[(modernData.L == mostLosses)]
print("The team with the worst record in the history of the MLB post WWI is:")
print (worstTeams)

The team with the most wins in the history post WWI is:
                  name  yearID lgID teamID franchID divID  Rank    G    W   L  \
2379  Seattle Mariners    2001   AL    SEA      SEA     W     1  162  116  46   

      ...   ERA WHIP IPouts    HA  HRA  BBA   SOA   E   DP     FP  
2379  ...  3.54  1.2   4395  1293  160  465  1051  83  137  0.986  

[1 rows x 36 columns]


The team with the worst record in the history of the MLB post WWI is:
               name  yearID lgID teamID franchID divID  Rank    G   W    L  \
1391  New York Mets    1962   NL    NYN      NYM   NaN    10  161  40  120   

      ...   ERA   WHIP IPouts    HA  HRA  BBA  SOA    E   DP     FP  
1391  ...  5.04  1.502   4290  1577  192  571  772  210  167  0.967  

[1 rows x 36 columns]


In [28]:
#Most losses and worst team

mostLosses = modernData.L.max()
worstTeams = modernData[(modernData.L == mostLosses)]
print("The team with the worst record in the history of the MLB post WWI is:")
print (worstTeams)

The team with the worst record in the history of the MLB post WWI is:
               name  yearID lgID teamID franchID divID  Rank    G   W    L  \
1391  New York Mets    1962   NL    NYN      NYM   NaN    10  161  40  120   

      ...   ERA   WHIP IPouts    HA  HRA  BBA  SOA    E   DP     FP  
1391  ...  5.04  1.502   4290  1577  192  571  772  210  167  0.967  

[1 rows x 36 columns]


In [78]:
#link championships to franchises (print just the y column)
champslist= modernData[modernData.WSWin == "Y"]
print("The number of times each franchise has won the World Series (championship of baseball)")
print(champslist.groupby("name")['WSWin'].count().reset_index().sort_values(by='WSWin',ascending=False))

The number of times each franchise has won the World Series (championship of baseball)
                      name  WSWin
19        New York Yankees     27
25     St. Louis Cardinals     11
14     Los Angeles Dodgers      6
20       Oakland Athletics      4
4           Boston Red Sox      4
8          Cincinnati Reds      4
17         New York Giants      4
10          Detroit Tigers      4
23      Pittsburgh Pirates      4
24    San Francisco Giants      3
3        Baltimore Orioles      3
18           New York Mets      2
16         Minnesota Twins      2
11         Florida Marlins      2
13      Kansas City Royals      2
12          Houston Astros      2
22   Philadelphia Phillies      2
9        Cleveland Indians      2
26       Toronto Blue Jays      2
2           Atlanta Braves      2
21  Philadelphia Athletics      2
27    Washington Nationals      1
0           Anaheim Angels      1
15        Milwaukee Braves      1
1     Arizona Diamondbacks      1
7        Chicago White Sox   

In [75]:
#link League championships to franchises (print just Y)
lgwins = modernData[modernData.LgWin == "Y"]
print("The number of times each franchise has won their respective league")
print(lgwins.groupby("name")['LgWin'].count().reset_index().sort_values(by='LgWin',ascending=False))

The number of times each franchise has won their respective league
                      name  LgWin
23        New York Yankees     40
31     St. Louis Cardinals     19
17     Los Angeles Dodgers     12
21         New York Giants      9
13          Detroit Tigers      8
10         Cincinnati Reds      8
5           Boston Red Sox      8
6         Brooklyn Dodgers      7
26   Philadelphia Phillies      7
8             Chicago Cubs      6
11       Cleveland Indians      6
24       Oakland Athletics      6
29    San Francisco Giants      6
3        Baltimore Orioles      6
2           Atlanta Braves      6
22           New York Mets      5
15          Houston Astros      5
27      Pittsburgh Pirates      5
16      Kansas City Royals      4
20         Minnesota Twins      3
36     Washington Senators      3
25  Philadelphia Athletics      3
28        San Diego Padres      2
32          Tampa Bay Rays      2
33           Texas Rangers      2
34       Toronto Blue Jays      2
18        Milwa

In [77]:
#link Division Winners to franchises (print just Y)
divwins = modernData[modernData.DivWin == "Y"]
print("The number of times each individual franchise has won their respective division")
print(divwins.groupby("name")['DivWin'].count().reset_index().sort_values(by='DivWin',ascending=False))

The number of times each individual franchise has won their respective division
                             name  DivWin
2                  Atlanta Braves      22
15            Los Angeles Dodgers      20
20               New York Yankees      20
21              Oakland Athletics      17
27            St. Louis Cardinals      15
22          Philadelphia Phillies      12
17                Minnesota Twins      12
12                 Houston Astros      12
4                  Boston Red Sox      10
8                 Cincinnati Reds      10
10              Cleveland Indians      10
25           San Francisco Giants       9
3               Baltimore Orioles       9
23             Pittsburgh Pirates       9
13             Kansas City Royals       8
6                    Chicago Cubs       8
11                 Detroit Tigers       7
29                  Texas Rangers       7
19                  New York Mets       6
7               Chicago White Sox       6
30              Toronto Blue Jays     

Generally speaking, the higher the counting stats a team has in batting (and therefore the higher composite stats), the better that team is at the offensive portion of the game. Conversely, the lower the counting stats a team has in pitching, the better their staff is at defense.

Below is a presentation of the best and worst offensive and defensive (respectively) metrics by a single team in the modern era of baseball

In [70]:
#OPS MAX
bestOPS = modernData.OPS.max()
bestBatting = modernData[modernData.OPS == bestOPS]
print("The team with the best composite batting statistics is:")
print(bestBatting)

#OPS MIN
worstOPS = modernData.OPS.min()
worstBatting = modernData[modernData.OPS == worstOPS]
print("The team with the worst composite batting statistics is:")
print(worstBatting)

The team with the best composite batting statistics is:
                 name  yearID lgID teamID franchID divID  Rank    G    W   L  \
824  New York Yankees    1927   AL    NYA      NYY   NaN     1  155  110  44   

     ...  ERA   WHIP IPouts    HA  HRA  BBA  SOA    E   DP     FP  
824  ...  3.2  1.305   4167  1403   42  409  431  196  123  0.969  

[1 rows x 36 columns]
The team with the worst composite batting statistics is:
               name  yearID lgID teamID franchID divID  Rank    G   W    L  \
1612  Texas Rangers    1972   AL    TEX      TEX     W     6  154  54  100   

      ...   ERA   WHIP IPouts    HA  HRA  BBA  SOA    E   DP     FP  
1612  ...  3.53  1.361   4124  1258   92  613  868  166  147  0.972  

[1 rows x 36 columns]


In [74]:
#WHIP MIN
bestWHIP = modernData.WHIP.min()
bestPitching = modernData[modernData.WHIP == bestWHIP]
print("The team with the best composite pitching statistics is:")
print(bestPitching)

#WHIP MAX
worstWHIP = modernData.WHIP.max()
worstPitching = modernData[modernData.WHIP == worstWHIP]
print("The team with the worst composite pitching statistics is:")
print(worstPitching)

The team with the best composite pitching statistics is:
                     name  yearID lgID teamID franchID divID  Rank    G    W  \
2998  Los Angeles Dodgers    2022   NL    LAN      LAD     W     1  162  111   

       L  ...  ERA   WHIP IPouts    HA  HRA  BBA   SOA   E   DP     FP  
2998  51  ...  2.8  1.048   4354  1114  152  407  1465  83  120  0.986  

[1 rows x 36 columns]
The team with the worst composite pitching statistics is:
                      name  yearID lgID teamID franchID divID  Rank    G   W  \
874  Philadelphia Phillies    1930   NL    PHI      PHI   NaN     8  156  52   

       L  ...   ERA   WHIP IPouts    HA  HRA  BBA  SOA    E   DP     FP  
874  102  ...  6.71  1.848   4116  1993  142  543  384  239  169  0.962  

[1 rows x 36 columns]
